### PR TITLE
Make repeated maps stack-safe using AndThen

### PR DESCRIPTION
**Pull request by TimWSpence**

### CODE REVIEW

File: core/shared/src/test/scala/fs2/StreamSuite.scala [hunk 1]
```diff
@@ -445,6 +445,13 @@ class StreamSuite extends Fs2Suite {
 
       assertEquals(unfoldTree(1).flatten.take(10).toList, List.tabulate(10)(_ + 1))
     }
+
+    test("regression #2353 - stack safety of map") {
+      def loop(str: Stream[Pure, Int], i: Int): Stream[Pure, Int] =
```
Comment (kubukoz) [inline core/shared/src/test/scala/fs2/StreamSuite.scala:450]: let's add `@tailrec`

**Comment by mpilquist** [comment]
Awesome! We'll need mima exclusions but otherwise this looks good to go.

**Comment by TimWSpence** [comment]
@mpilquist are those exclusions ok? I'm actually wondering if we might be able to get away with keeping the field type as `Function1` anyway and just constructing `AndThen` in `mapOutput`. It's a subtype so should be ok

**Comment by mpilquist** [comment]
Yep they look good. Thanks!

**Pull request merged**